examples/mxnet_imagenet_resnet50.py

```diff
@@ -0,0 +1,337 @@
+        data = np.random.uniform(-1, 1, data_shape)
+        self.data = mx.nd.array(data, dtype=self.dtype,
+                                ctx=ctx)
+        self.label = mx.nd.array(label, dtype=self.dtype,
+                                 ctx=ctx)
+
+    def __iter__(self):
+        return self
+
+    @property
+    def provide_data(self):
+        return [mx.io.DataDesc('data', self.data.shape, self.dtype)]
+
+    @property
+    def provide_label(self):
+        return [mx.io.DataDesc('softmax_label',
+                               (self.batch_size,), self.dtype)]
+
+    def next(self):
+        self.cur_iter += 1
+        if self.cur_iter <= self.max_iter:
+            return DataBatch(data=(self.data,),
+                             label=(self.label,),
+                             pad=0,
+                             index=None,
+                             provide_data=self.provide_data,
+                             provide_label=self.provide_label)
+        else:
+            raise StopIteration
+
+    def __next__(self):
+        return self.next()
+
+    def reset(self):
+        self.cur_iter = 0
+
+if args.use_rec:
+    # Fetch training and validation data if present
+    train_data, val_data, batch_fn = get_data_rec(args.rec_train,
+                                                  args.rec_train_idx,
+                                                  args.rec_val,
+                                                  args.rec_val_idx,
+                                                  batch_size,
+                                                  args.data_nthreads)
+else:
+    # Otherwise use synthetic data
+    image_shape = (3, 224, 224)
+    data_shape = (batch_size,) + image_shape
+    train_data = SyntheticDataIter(num_classes, data_shape, epoch_size,
+                                   np.float32, context)
+    val_data = None
+
+
+def main():
+    # Get model from GluonCV model zoo
+    # https://gluon-cv.mxnet.io/model_zoo/index.html
+    net = get_model(args.model, **kwargs)
+    net.cast(args.dtype)
+
+    # Create input symbol
+    data = mx.sym.var('data')
+    if args.dtype == 'float16':
+        data = mx.sym.Cast(data=data, dtype=np.float16)
+        net.cast(np.float16)
+
+    # Create output symbol
+    out = net(data)
+    if args.dtype == 'float16':
+        out = mx.sym.Cast(data=out, dtype=np.float32)
+    softmax = mx.sym.SoftmaxOutput(out, name='softmax')
+
+    if args.use_pretrained:
+        arg_params = {}
+        for x in net.collect_params().values():
+            x.reset_ctx(mx.cpu())
+            arg_params[x.name] = x.data()
+    else:
+        arg_params = None
+    aux_params = None
+
+    # Create model
+    mod = mx.mod.Module(softmax, context=context)
+
+    # Create optimizer
+    optimizer_params = {'wd': args.wd,
+                        'momentum': args.momentum,
+                        'rescale_grad': 1.0 / batch_size,
```

Comment:
What does this rescaling accomplish?  Should LR be decreased to get the same effect?

Reply:
Rescale_grad is used to multiply the gradients before updating. When using MXNet Module API and calling [fit](https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/module/base_module.py#L440), you can give a name string or an Optimizer object for the optimizer. If a name string is given, the rescale_grad is normalized to 1.0 / batch_size [intermally](https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/module/module.py#L517) and users do not need to specify it in the optimizer_params.  However, in this example, we need to give an Optimizer object to the fit API because we need to create the optimizer first and wrap it with the DistributedOptimizer. In this case, we need to specify this rescale_grad when creating the [optimizer](https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/optimizer/optimizer.py#L46). 

Decreasing LR (to LR/batch_size) cannot get the same effect directly. I did one experiment by removing this rescale_grad and decreasing the LR. I got 69% accuracy in 90 epoches while the same training with rescale_grad specified achieved 76% accuracy. And specifying rescale_grad can also keep the same user experience for current MXNet users because they can use the same LR value they used in the Parameter Server approach with Horovod.

Reply:
I see, so `rescale_grad` is not Horovod-specific?

Reply:
It's not. Even without Horovod, we have the option to create Optimizer first (need to specify rescale_grad) and pass it to the fit API but we usually do not do that.